crypt_test.go

```diff
@@ -1597,9 +1747,234 @@ func (s *cryptSuite) TestActivateVolumeWithMultipleKeyData10(c *C) {
+		model:            models[0],
+		activateSlots:    []int{luks2.AnySlot},
+		validKey:         key,
+		validAuxKey:      auxKey})
+}
+
+func (s *cryptSuite) TestActivateVolumeWithMultipleKeyData16(c *C) {
+	// Test activation with orphaned LUKS token but valid external token
+	// prints "Cannot read keydata from token" to os.Stderr
+	token := &luksview.KeyDataToken{
+		TokenBase: luksview.TokenBase{
+			TokenKeyslot: 1,
+			TokenName:    "default",
+		},
+		Data: json.RawMessage("foo")}
+	s.addMockToken("/dev/sda1", token)
+
+	keyData, key, auxKey := s.newNamedKeyData(c, "")
+
+	models := []SnapModel{
+		testutil.MakeMockCore20ModelAssertion(c, map[string]interface{}{
+			"authority-id": "fake-brand",
+			"series":       "16",
+			"brand-id":     "fake-brand",
+			"model":        "fake-model",
+			"grade":        "secured",
+		}, "Jv8_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij")}
+	c.Check(keyData.SetAuthorizedSnapModels(auxKey, models...), IsNil)
+
+	stderr := new(bytes.Buffer)
+	restore := MockStderr(stderr)
+	defer restore()
+
+	s.testActivateVolumeWithMultipleKeyData(c, &testActivateVolumeWithMultipleKeyDataData{
+		keys:             []DiskUnlockKey{key},
+		keyData:          []*KeyData{keyData},
+		volumeName:       "data",
+		sourceDevicePath: "/dev/sda1",
+		model:            models[0],
+		activateSlots:    []int{luks2.AnySlot},
+		validKey:         key,
+		validAuxKey:      auxKey})
+	c.Check(stderr.String(), Equals, "secboot: Cannot read keydata from token default: cannot decode key data: invalid character 'o' in literal false (expecting 'a')\n")
+}
+
+func (s *cryptSuite) TestActivateVolumeWithMultipleKeyData17(c *C) {
+	// Test activation with invalid (containing an invalid key) and valid LUKS token.
+	s.addMockKeyslot("/dev/sda1", nil) // add an empty slot for the invalid token
+
+	keyData, keys, auxKeys := s.newMultipleNamedKeyData(c, "", "")
+
+	models := []SnapModel{
+		testutil.MakeMockCore20ModelAssertion(c, map[string]interface{}{
+			"authority-id": "fake-brand",
+			"series":       "16",
+			"brand-id":     "fake-brand",
+			"model":        "fake-model",
+			"grade":        "secured",
+		}, "Jv8_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij")}
+	c.Check(keyData[0].SetAuthorizedSnapModels(auxKeys[0], models...), IsNil)
+	c.Check(keyData[1].SetAuthorizedSnapModels(auxKeys[1], models...), IsNil)
+
+	for i, kd := range keyData {
+		w := makeMockKeyDataWriter()
+		c.Check(kd.WriteAtomic(w), IsNil)
+
+		token := &luksview.KeyDataToken{
+			TokenBase: luksview.TokenBase{
+				TokenName:    fmt.Sprintf("default%d", i),
+				TokenKeyslot: i,
+			},
+			Data:     w.final.Bytes(),
+			Priority: 0}
+		s.addMockToken("/dev/sda1", token)
+	}
+
+	s.testActivateVolumeWithMultipleKeyData(c, &testActivateVolumeWithMultipleKeyDataData{
+		keys:             keys[1:],
+		volumeName:       "data",
+		sourceDevicePath: "/dev/sda1",
+		model:            models[0],
+		activateSlots:    []int{0, 1},
```

Comment:
I don't see a test that has both the trying of provided keys (failing) first and the slots later...

Reply:
Isn't that case covered by `TestActivateVolumeWithMultipleKeyData14`?

Reply:
In `TestActivateVolumeWithMultipleKeyData14`, the externally provided key fails because it fails the snap model authorization check. I've added an additional test where the externally provided key fails activation as well.